dali/kernels/tensor_view.h

```diff
@@ -0,0 +1,237 @@
+// Copyright (c) 2017-2018, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_TENSOR_VIEW_H_
+#define DALI_KERNELS_TENSOR_VIEW_H_
+
+#include "shape.h"
+
+namespace tensor {
+
+struct EmptyBackendTag {};
+
+template <typename Backend, typename DataType, int dim_>
+struct TensorView;
+
+template <typename Backend, typename DataType>
+struct TensorView<Backend, DataType, DynamicTensorShape> {
+  TensorView() : data(nullptr), shape() {}
+
+  template <int dim>
+  TensorView(DataType *data, TensorShape<dim> shape) : data(data), shape(shape) {}
+  TensorView(const TensorView &) = default;
+  TensorView(TensorView &&) = default;
+  TensorView &operator=(const TensorView &) = default;
+  TensorView &operator=(TensorView &&) = default;
+
+  // Dynamic accepts anything
+  template <int other_dim>
+  TensorView(const TensorView<Backend, DataType, other_dim> &other)
+      : data(other.data), shape(other.shape) {}
+
+  template <int other_dim>
+  TensorView(TensorView<Backend, DataType, other_dim> &&other)
+      : data(other.data), shape(other.shape) {}
+
+  template <int other_dim>
+  TensorView &operator=(const TensorView<Backend, DataType, other_dim> &other) {
+    data = other.data;
+    shape = other.shape;
+    return *this;
+  }
+
+  template <int other_dim>
+  TensorView &operator=(TensorView<Backend, DataType, other_dim> &&other) {
+    data = std::move(other.data);  // TODO(klecki), should I null it?
```

Comment:
There should be a move assignment and construction only for dynamic `other`. This variant will destroy `other`'s shape and, as such should also nullify the pointer. Same applies to the constructor.
When `other` has static dim, we have no reason to even provide a move - it has no advantages over copy, so we'd just make a copy. Even if someone uses it with `std::move`, R-value reference will decay to const L-value reference and we're set.

Reply:
Should be done, if I didn't miss any.